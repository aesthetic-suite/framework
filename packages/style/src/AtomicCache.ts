import { CacheItem, StyleParams } from './types';

// https://jsperf.com/javascript-objects-vs-map-performance
// https://jsperf.com/performance-of-map-vs-object
export default class AtomicCache {
  cache: { [property: string]: { [value: string]: CacheItem[] } } = {};

  match(item: CacheItem, params: StyleParams, minimumRank?: number): boolean {
    if (minimumRank !== undefined && item.rank < minimumRank) {
      return false;
    }

    if (item.selector !== params.selector) {
      return false;
    }

    if (item.type !== params.type) {
      return false;
    }

    if (Array.isArray(params.conditions) && params.conditions.length > 0) {
      return params.conditions.every(i =>
        item.conditions.find(p => p.query === i.query && p.type === i.type),
      );
    }

    return true;
  }

  read(
    property: string,
    value: string,
    params: StyleParams,
    minimumRank?: number,
  ): CacheItem | null {
    const propertyCache = this.cache[property];

    if (!propertyCache) {
      return null;
    }

    const valueCache = propertyCache[value];

    if (!valueCache || valueCache.length === 0) {
      return null;
    }

    return valueCache.find(item => this.match(item, params, minimumRank)) ?? null;
  }

  write(property: string, value: string, item: CacheItem): this {
    let propertyCache = this.cache[property];

    if (!propertyCache) {
      propertyCache = {};
      this.cache[property] = propertyCache;
    }

    let valueCache = propertyCache[value];

    if (!valueCache) {
      valueCache = [];
      propertyCache[value] = valueCache;
    }

    valueCache.push(item);

    return this;
  }
}
