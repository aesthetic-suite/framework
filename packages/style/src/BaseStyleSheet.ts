import { isSSR } from '@aesthetic/utils';
import { StyleRule } from './types';

export default abstract class BaseStyleSheet {
  lastIndex: number = -1;

  sheet: StyleRule;

  protected rafHandle: number = 0;

  protected ruleBuffer: { index: number; rule: string }[] = [];

  constructor(sheet: StyleRule) {
    this.sheet = sheet;
  }

  protected enqueueRule(rule: string, customIndex?: number): number {
    let index = 0;

    // When testing or SSR, just insert the rule immediately
    if (process.env.NODE_ENV === 'test' || isSSR()) {
      index = customIndex ?? this.sheet.cssRules.length;

      this.injectRule(rule, index);

      // Otherwise buffer the insert until the next animation frame
    } else {
      index = customIndex ?? (this.lastIndex += 1);

      this.ruleBuffer.push({
        index,
        rule,
      });

      if (!this.rafHandle) {
        this.rafHandle = requestAnimationFrame(this.flushRules);
      }
    }

    return index;
  }

  protected injectRule = (rule: string, index: number) => {
    // CSS is locked in developer tools when using `insertRule`,
    // so avoid using it in development so we have full control.
    // istanbul ignore next
    if (process.env.NODE_ENV === 'development' && !isSSR()) {
      this.sheet.textContent += rule;
    } else {
      this.sheet.insertRule(rule, index);
      this.lastIndex = this.sheet.cssRules.length;
    }
  };

  protected flushRules = () => {
    this.rafHandle = 0;
    this.ruleBuffer.forEach((buffer) => {
      this.injectRule(buffer.rule, buffer.index);
    });
  };
}
