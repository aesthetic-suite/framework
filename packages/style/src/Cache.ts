import { Value, ValueWithFallbacks } from '@aesthetic/types';
import { CacheManager, CacheItem, CacheStorage, RenderOptions } from './types';

export function createCacheKey(
  property: string,
  value: Value | ValueWithFallbacks,
  { selector, conditions }: RenderOptions,
): string {
  let key = `${property}:${value}`;

  if (selector) {
    key += `:${selector}`;
  }

  if (conditions && conditions.length > 0) {
    key += `:${conditions.join(':')}`;
  }

  return key;
}

export function createCacheManager(defaultItems: CacheStorage = {}): CacheManager {
  const cache: CacheStorage = defaultItems;

  return {
    read(key, minimumRank) {
      const items = cache[key];

      if (!items) {
        return null;
      } else if (minimumRank === undefined) {
        return items[0];
      }

      return items.find((item) => item.rank! >= minimumRank) || null;
    },

    write(key: string, item: CacheItem) {
      const result = cache[key] || [];

      result.push(item);

      cache[key] = result;
    },
  };
}
