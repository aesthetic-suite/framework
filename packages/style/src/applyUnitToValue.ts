import { isUnitlessProperty } from 'css-in-js-utils';
import { Value } from './types';

export default function applyUnitToValue(property: string, value: Value): string {
  if (typeof value === 'string' || isUnitlessProperty(property) || value === 0) {
    return String(value);
  }

  return `${value}px`;
}
