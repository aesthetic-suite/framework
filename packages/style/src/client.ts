import { Variables } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import createStyleElements, { getStyleElement } from './client/dom';
import hydrateStyles from './client/hydration';
// Rollup compatibility
import {
  createCacheManager,
  createStyleEngine,
  createSheetManager,
  formatVariable,
  EngineOptions,
  StyleEngine,
} from './index';

export { createStyleElements, getStyleElement };

function setRootVariables(vars: Variables) {
  const root = document.documentElement;

  objectLoop(vars, (value, key) => {
    root.style.setProperty(formatVariable(key), String(value));
  });
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  const engine: StyleEngine = {
    ...createStyleEngine({
      cacheManager: createCacheManager(),
      sheetManager: createSheetManager(createStyleElements()),
      ...options,
    }),
    setRootVariables,
  };

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
