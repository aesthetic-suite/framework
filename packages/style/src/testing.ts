/* eslint-disable unicorn/import-index */

import { CSS } from '@aesthetic/types';
import {
  CacheManager,
  StyleEngine,
  EngineOptions,
  SheetManager,
  SheetType,
  StyleRule,
  getDocumentStyleSheet,
} from './index';
import { createCacheManager, createStyleElements, createEngine, createSheetManager } from './next';

export function createTestCacheManager(): CacheManager {
  return createCacheManager();
}

export function createTestSheetManager(): SheetManager {
  return createSheetManager(createStyleElements());
}

export function createTestStyleEngine(options: Partial<EngineOptions>): StyleEngine {
  return createEngine({
    cacheManager: createTestCacheManager(),
    sheetManager: createTestSheetManager(),
    ...options,
  });
}

export function getRenderedStyles(type: SheetType | StyleRule): CSS {
  return Array.from(
    ((typeof type === 'string' ? getDocumentStyleSheet(type) : type) as StyleRule).cssRules,
  ).reduce((css, rule) => css + rule.cssText, '');
}

export function purgeStyles(type?: SheetType): void {
  if (type) {
    // This is the only way to generate accurate snapshots.
    // It may slow down tests though?
    document.getElementById(`aesthetic-${type}`)?.remove();
  } else {
    purgeStyles('global');
    purgeStyles('standard');
    purgeStyles('conditions');
  }
}
