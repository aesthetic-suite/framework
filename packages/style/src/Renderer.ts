import hash from 'string-hash';
import { isObject } from 'aesthetic-utils';
import { hyphenateProperty } from 'css-in-js-utils';
import AtomicCache from './AtomicCache';
import { ClassName, Properties, StyleParams, Property, Value } from './types';
import getDocumentStyleSheet from './getDocumentStyleSheet';
import isMediaQueryCondition from './isMediaQueryCondition';
import isSupportsCondition from './isSupportsCondition';
import isNestedSelector from './isNestedSelector';

export default class Renderer {
  protected atRuleCache = new Map();

  protected classNameCache = new AtomicCache();

  constructor() {
    // Create the elements in a strict order
    getDocumentStyleSheet('global');
    getDocumentStyleSheet('low-pri');
    getDocumentStyleSheet('high-pri');
  }

  render(properties: Properties, params: StyleParams): ClassName {
    return Object.entries(properties)
      .map(([key, value]) => {
        if (value === null || value === undefined) {
          return '';
        }

        if (isObject(value)) {
          if (isMediaQueryCondition(key) || isSupportsCondition(key)) {
            return this.render(value, { ...params, condition: key });
          } else if (isNestedSelector(key)) {
            return this.render(value, { ...params, selector: key });
          } else if (__DEV__) {
            console.warn(`Unknown property "${key}" with value "${value}".`);
          }

          return '';
        }

        return this.renderDeclaration(key as Property, value, params);
      })
      .join(' ');
  }

  renderDeclaration<K extends Property>(property: K, value: Properties[K], params: StyleParams) {
    const key = hyphenateProperty(property);
    const item = this.classNameCache.read(key, value, params);

    if (item) {
      return item.className;
    }

    const sheet = getDocumentStyleSheet(params.type);
    const rank = sheet.cssRules.length;
    const className = this.createUniqueHash(property, value, params);

    sheet.insertRule(this.createRule(className, key, value, params), rank);

    this.classNameCache.write(key, value, {
      condition: '',
      selector: '',
      type: 'low-pri',
      ...params,
      className,
      rank,
    });

    return className;
  }

  renderFontFace() {}

  renderKeyframes() {}

  protected createDeclaration(property: string, value: Value): string {
    return `${property}: ${value}`;
  }

  protected createDeclarationBlock(properties: Properties): string {
    return Object.entries(properties)
      .map(([prop, value]) => this.createDeclaration(prop, value))
      .join(';');
  }

  protected createRule(
    className: ClassName,
    property: string,
    value: Value,
    params: StyleParams,
  ): string {
    let rule = `.${className}`;

    if (params.selector) {
      rule += params.selector;
    }

    rule += `{ ${this.createDeclaration(property, value)} }`;

    if (params.condition) {
      rule = `${params.condition} { ${rule} }`;
    }

    return rule;
  }

  protected createUniqueHash(property: string, value: Value, params: StyleParams): string {
    return hash(`${params.condition || ''}${params.selector || ''}${property}${value}`);
  }
}
