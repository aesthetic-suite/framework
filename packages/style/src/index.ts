/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Direction, EngineOptions, VariablesMap } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { createStyleElements, getStyleElement } from './client/dom';
import { hydrateStyles } from './client/hydration';
import { createCacheManager } from './common/cache';
import { createStyleEngine } from './common/engine';
import { createSheetManager } from './common/sheet';
import { formatVariable } from './common/syntax';
import { StyleEngine } from './types';

export * from './common/cache';
export * from './common/constants';
export * from './common/engine';
export * from './common/helpers';
export * from './common/sheet';
export * from './common/syntax';
export * from './types';

export { createStyleElements, getStyleElement };

// Set active direction on the document `dir` attribute
function setDirection(direction: Direction) {
  document.documentElement.setAttribute('dir', direction);
}

// Set CSS variables to :root
function setRootVariables(vars: VariablesMap) {
  objectLoop(vars, (value, key) => {
    document.documentElement.style.setProperty(formatVariable(key), String(value));
  });
}

// Set active theme class names on the `body`
function setTheme(classNames: string[]) {
  document.body.className = classNames.join(' ');
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  if (global.AESTHETIC_CUSTOM_ENGINE) {
    return global.AESTHETIC_CUSTOM_ENGINE as StyleEngine;
  }

  const direction = (document.documentElement.getAttribute('dir') ||
    document.body.getAttribute('dir') ||
    'ltr') as Direction;

  const engine: StyleEngine = createStyleEngine({
    cacheManager: createCacheManager(),
    direction,
    sheetManager: createSheetManager(createStyleElements()),
    ...options,
  });

  // Match against browser preferences
  engine.prefersColorScheme = (scheme) => matchMedia(`(prefers-color-scheme: ${scheme})`).matches;
  engine.prefersContrastLevel = (level) => matchMedia(`(prefers-contrast: ${level})`).matches;

  // Handle DOM specific logic
  engine.setDirection = setDirection;
  engine.setRootVariables = setRootVariables;
  engine.setTheme = setTheme;

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
