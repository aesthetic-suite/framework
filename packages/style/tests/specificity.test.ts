import Renderer from '../src/client/ClientRenderer';
import { getRenderedStyles, purgeStyles } from '../src/testing';

describe('Specificity', () => {
  let renderer: Renderer;

  beforeEach(() => {
    renderer = new Renderer();
  });

  afterEach(() => {
    purgeStyles('global');
    purgeStyles('standard');
    purgeStyles('conditions');
  });

  it('inserts declarations in the order they are defined', () => {
    renderer.renderRule({
      margin: 0,
      padding: '1px',
      width: '50px',
    });

    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('inserts declarations in the order they are defined (reversed)', () => {
    renderer.renderRule({
      width: '50px',
      padding: '1px',
      margin: 0,
    });

    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('inserts selectors in the order they are defined', () => {
    renderer.renderRule({
      color: 'white',
      ':active': {
        color: 'red',
      },
      ':hover': {
        color: 'blue',
      },
    });

    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  it('inserts selectors in the order they are defined (reversed)', () => {
    renderer.renderRule({
      color: 'white',
      ':hover': {
        color: 'blue',
      },
      ':active': {
        color: 'red',
      },
    });

    expect(getRenderedStyles('standard')).toMatchSnapshot();
  });

  describe('rule sets', () => {
    const ruleSet = {
      button: {
        padding: '8px',
        display: 'inline-block',
        color: 'red',

        ':hover': {
          color: 'darkred',
        },
      },
      buttonActive: {
        color: 'darkred',
      },
      buttonDisabled: {
        color: 'gray',
      },
    };

    it('renders in defined order if no explicit order provided', () => {
      const className = renderer.renderRuleSets(ruleSet);

      expect(className).toBe('a b c d e f');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('renders in a explicit order', () => {
      const className = renderer.renderRuleSets(ruleSet, [
        'buttonActive',
        'buttonDisabled',
        'button',
      ]);

      expect(className).toBe('a b c d e f');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('renders the same class names from previous render', () => {
      const classNameA = renderer.renderRuleSets(ruleSet, ['buttonActive']);
      const classNameB = renderer.renderRuleSets(ruleSet, ['buttonActive']);

      expect(classNameA).toBe('a');
      expect(classNameA).toBe(classNameB);
      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('can omit sets', () => {
      const className = renderer.renderRuleSets(ruleSet, ['buttonActive']);

      expect(className).toBe('a');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });

    it('can render set by set', () => {
      const a = renderer.renderRuleSets(ruleSet, ['buttonActive']);

      expect(a).toBe('a');
      expect(getRenderedStyles('standard')).toMatchSnapshot();

      const b = renderer.renderRuleSets(ruleSet, ['button']);

      expect(b).toBe('b c d e');
      expect(getRenderedStyles('standard')).toMatchSnapshot();

      const c = renderer.renderRuleSets(ruleSet, ['buttonDisabled']);

      expect(c).toBe('f');
      expect(getRenderedStyles('standard')).toMatchSnapshot();
    });
  });
});
