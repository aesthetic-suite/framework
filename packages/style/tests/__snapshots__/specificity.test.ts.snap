// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Specificity inserts declarations in the order they are defined (reversed) 1`] = `".a {width: 50px;}.b {padding: 1px;}.c {margin: 0;}"`;

exports[`Specificity inserts declarations in the order they are defined 1`] = `".a {margin: 0;}.b {padding: 1px;}.c {width: 50px;}"`;

exports[`Specificity inserts selectors in the order they are defined (reversed) 1`] = `".a {color: white;}.b:hover {color: blue;}.c:active {color: red;}"`;

exports[`Specificity inserts selectors in the order they are defined 1`] = `".a {color: white;}.b:active {color: red;}.c:hover {color: blue;}"`;

exports[`Specificity ordered rules can omit sets 1`] = `".a {color: darkred;}"`;

exports[`Specificity ordered rules can render set by set 1`] = `".a {color: darkred;}"`;

exports[`Specificity ordered rules can render set by set 2`] = `".a {color: darkred;}.b {padding: 8px;}.c {display: inline-block;}.d {color: red;}.e:hover {color: darkred;}"`;

exports[`Specificity ordered rules can render set by set 3`] = `".a {color: darkred;}.b {padding: 8px;}.c {display: inline-block;}.d {color: red;}.e:hover {color: darkred;}.f {color: gray;}"`;

exports[`Specificity ordered rules renders in a explicit order 1`] = `".a {color: darkred;}.b {color: gray;}.c {padding: 8px;}.d {display: inline-block;}.e {color: red;}.f:hover {color: darkred;}"`;

exports[`Specificity ordered rules renders in defined order if no explicit order provided 1`] = `".a {padding: 8px;}.b {display: inline-block;}.c {color: red;}.d:hover {color: darkred;}.e {color: darkred;}.f {color: gray;}"`;

exports[`Specificity ordered rules renders the same class names from previous render 1`] = `".a {color: darkred;}"`;
